include/camp/tuple.hpp

```diff
@@ -0,0 +1,253 @@
+  }
+};
+
+template <camp::idx_t i, typename T>
+struct tuple_element;
+template <camp::idx_t i, typename... Types>
+struct tuple_element<i, tuple<Types...>> {
+  using type = camp::at_v<typename tuple<Types...>::TList, i>;
+};
+template <camp::idx_t i, typename T>
+using tuple_element_t = typename tuple_element<i, T>::type;
+
+template <int index, typename... Args>
+CAMP_HOST_DEVICE constexpr auto get(const tuple<Args...>& t) noexcept
+    -> tpl_get_ret<tuple<Args...>, index> const&
+{
+  static_assert(sizeof...(Args) > index, "index out of range");
+  return t.tpl_get_store<tuple<Args...>, index>::get_inner();
+}
+
+template <int index, typename... Args>
+CAMP_HOST_DEVICE constexpr auto get(tuple<Args...>& t) noexcept
+    -> tpl_get_ret<tuple<Args...>, index>&
+{
+  static_assert(sizeof...(Args) > index, "index out of range");
+  return t.tpl_get_store<tuple<Args...>, index>::get_inner();
+}
+
+template <typename Tuple>
+struct tuple_size;
+
+template <typename... Args>
+struct tuple_size<tuple<Args...>> {
+  static constexpr size_t value = sizeof...(Args);
+};
+
+template <typename... Args>
+struct tuple_size<tuple<Args...>&> {
+  static constexpr size_t value = sizeof...(Args);
+};
+
+template <typename... Args>
+CAMP_HOST_DEVICE constexpr auto make_tuple(Args&&... args)
+    -> tuple<internal::special_decay_t<Args>...>
+{
+  return tuple<internal::special_decay_t<Args>...>{std::forward<Args>(args)...};
+}
+
+template <typename... Args>
+CAMP_HOST_DEVICE constexpr auto forward_as_tuple(Args&&... args) noexcept
+    -> tuple<Args&&...>
+{
+  return tuple<Args&&...>(std::forward<Args>(args)...);
+}
+
+template <class... Types>
+CAMP_HOST_DEVICE constexpr tuple<Types&...> tie(Types&... args) noexcept
+{
+  return tuple<Types&...>{args...};
+}
+
+template <typename... Lelem,
+          typename... Relem,
+          camp::idx_t... Lidx,
+          camp::idx_t... Ridx>
+CAMP_HOST_DEVICE constexpr auto tuple_cat_pair(tuple<Lelem...>&& l,
+                                               camp::idx_seq<Lidx...>,
+                                               tuple<Relem...>&& r,
+                                               camp::idx_seq<Ridx...>) noexcept
+    -> tuple<Lelem..., Relem...>
+{
+  return make_tuple(get<Lidx>(l)..., get<Ridx>(r)...);
+}
+
+template <typename Fn, camp::idx_t... Sequence, typename TupleLike>
+CAMP_HOST_DEVICE constexpr auto invoke_with_order(TupleLike&& t,
+                                                  Fn&& f,
+                                                  camp::idx_seq<Sequence...>)
+    -> decltype(f(get<Sequence>(t)...))
+{
+  return f(get<Sequence>(t)...);
+}
+
+template <typename Fn, typename TupleLike>
+CAMP_HOST_DEVICE constexpr auto invoke(TupleLike&& t, Fn&& f) -> decltype(
+    invoke_with_order(t,
+                      f,
+                      camp::make_idx_seq_t<tuple_size<TupleLike>::value>{}))
+{
+  return invoke_with_order(
+      t, f, camp::make_idx_seq_t<tuple_size<TupleLike>::value>{});
```

Comment:
Universal references should be perfectly forwarded

Reply:
Thanks for the catch on this, not sure how I missed it until now but this is fixed.